dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/trackedentity/DefaultTrackerAccessManager.java

```diff
@@ -59,31 +63,93 @@ public class DefaultTrackerAccessManager implements TrackerAccessManager {
 
   private final AclService aclService;
   private final TrackerOwnershipManager ownershipAccessManager;
-
+  private final ProgramService programService;
+
+  /**
+   * Check the data read permissions and ownership of a tracked entity given the programs for which
+   * the user has metadata access to.
+   *
+   * @return No errors if a user has access to at least one program
+   */
   @Override
   public List<String> canRead(UserDetails user, TrackedEntity trackedEntity) {
     // always allow if user == null (internal process) or user is superuser
     if (user == null || user.isSuper() || trackedEntity == null) {
       return List.of();
     }
 
-    OrganisationUnit ou = trackedEntity.getOrganisationUnit();
-    List<String> errors = new ArrayList<>();
-    // ou should never be null, but needs to be checked for legacy reasons
-    if (ou != null && !user.isInUserSearchHierarchy(ou.getPath())) {
-      errors.add(NO_READ_ACCESS_TO_ORG_UNIT + ": " + ou.getUid());
+    return new ArrayList<>(canRead(user, trackedEntity, programService.getAllPrograms()));
+  }
+
+  private Set<String> canRead(
+      UserDetails user, TrackedEntity trackedEntity, List<Program> programs) {
+
+    if (null == trackedEntity) {
+      return Set.of();
     }
 
     TrackedEntityType trackedEntityType = trackedEntity.getTrackedEntityType();
 
     if (!aclService.canDataRead(user, trackedEntityType)) {
-      errors.add(
+      return Set.of(
           "User has no data read access to tracked entity type: " + trackedEntityType.getUid());
     }
 
+    initializeTrackedEntityOrgUnitParents(trackedEntity);
+
+    Set<String> errors = new HashSet<>();
+
+    List<Program> tetPrograms =
+        programs.stream()
+            .filter(
+                p -> Objects.equals(p.getTrackedEntityType(), trackedEntity.getTrackedEntityType()))
+            .toList();
+
+    if (tetPrograms.isEmpty()) {
+      return Set.of("User has no access to any program");
+    }
+
+    for (Program program : tetPrograms) {
+      List<String> e = canRead(user, trackedEntity, program);
+      if (e.isEmpty()) {
+        return Set.of();
+      } else {
+        errors.addAll(e);
```

Comment:
Shouldn't we just say that user cannot access TE?
I can see that here there could be a lot of info that are not interesting for the user, potentially all the programs for the same TET that the user has no access.

Reply:
These errors are not shown to the user, now they are converted to a forbidden exception or will filter out relationships like in this case. I understand @muilpp wants to move to use error codes which will then help to move to a not found / forbidden logic as a response so I'll leave this job for later